tokio/src/runtime/task/list.rs

```diff
@@ -111,25 +111,27 @@ impl<S: 'static> OwnedTasks<S> {
             // to the field.
             task.header().set_owner_id(self.id);
         }
-
-        let mut lock = self.inner.lock();
-        if lock.closed {
-            drop(lock);
-            drop(notified);
+        // check close flag
+        if self.closed.load(Ordering::Acquire) {
             task.shutdown();
-            None
-        } else {
-            lock.list.push_front(task);
-            Some(notified)
+            return None;
         }
+        self.push_inner(id, task);
+        Some(notified)
+    }
+
+    #[inline]
+    fn push_inner(&self, task_id: super::Id, task: Task<S>) {
+        let mut lock = self.lists[task_id.0 as usize & (self.grain - 1)].lock();
+        lock.push_front(task);
+        self.count.fetch_add(1, Ordering::Relaxed);
     }
```

Comment:
This doesn't look right to me. The `OwnedTasks` could be closed between the call to `self.closed.load(Acquire)` and the call to `self.lists[...].lock()`. If that happens, then we push something to the list, but it's never removed because the close operation has already completed.

Reply:
You must check the `closed` flag _after_ calling `lock()` to fix this.

Reply:
No, I don't agree that this is correct. If I spawn something from outside the runtime during shutdown, this problem can result in the task being pushed to the `OwnedTasks`, but not removed by the `close_and_shutdown_all` call.

Reply:
Does the `Runtime.enter()` will cause such concurrency problem?

Reply:
You can spawn tasks during shutdown with `tokio::runtime::Handle`.

However, it doesn't really matter. Even if the API prevents you from triggering this, we should still fix it. Having correctness rely on such things makes it very easy to introduce bugs when changing things.

Reply:
In this cases, If a thread entered the runtime, then it become a legal worker thread.

This thread shoud be responsible for cleaning tasks, when the`EnterGuard` is dropped, let it check if the runtime is closed, if it is true, then cleaning tasks.

Does it make sense?

Reply:
> You can spawn tasks during shutdown with `tokio::runtime::Handle`.
> 
> However, it doesn't really matter. Even if the API prevents you from triggering this, we should still fix it. Having correctness rely on such things makes it very easy to introduce bugs when changing things.

I just saw this. What I said above may be a solution, but it may be more appropriate to directly use locks to solve the problem. I think your solution is more reasonable, let it easy.

Reply:
I see you moved the check into the lock, so I'll resolve this.